File: backend/m2/src/keycloak/keycloak_admin.js
const axios = require('axios');

const config = {
    baseUrl: process.env.KEYCLOAK_INTERNAL_URL || process.env.KEYCLOAK_URL || 'http://localhost:8091',
    realm: process.env.KEYCLOAK_REALM || 'Universidad',
    adminRealm: process.env.KEYCLOAK_ADMIN_REALM || 'master',
    adminUser: process.env.KEYCLOAK_ADMIN_USER || 'admin',
    adminPass: process.env.KEYCLOAK_ADMIN_PASSWORD || '123456',
    adminClient: process.env.KEYCLOAK_ADMIN_CLIENT || 'admin-cli'
};

async function getAdminToken() {
    const params = new URLSearchParams();
    params.append('grant_type', 'password');
    params.append('client_id', config.adminClient);
    params.append('username', config.adminUser);
    params.append('password', config.adminPass);

    const response = await axios.post(
        `${config.baseUrl}/realms/${config.adminRealm}/protocol/openid-connect/token`,
        params,
        { headers: { 'Content-Type': 'application/x-www-form-urlencoded' } }
    );

    return response.data.access_token;
}

async function findUserId(token, username) {
    const response = await axios.get(
        `${config.baseUrl}/admin/realms/${config.realm}/users`,
        {
            params: { username, exact: true },
            headers: { Authorization: `Bearer ${token}` }
        }
    );

    return response.data[0]?.id;
}

async function createOrGetUser(token, user) {
    const payload = {
        username: user.usuario,
        firstName: user.nombre,
        email: user.correo || undefined,
        enabled: true,
        emailVerified: true,
        credentials: [
            {
                type: 'password',
                value: user.pass,
                temporary: false
            }
        ]
    };

    try {
        await axios.post(
            `${config.baseUrl}/admin/realms/${config.realm}/users`,
            payload,
            { headers: { Authorization: `Bearer ${token}` } }
        );
    } catch (error) {
        if (error.response?.status !== 409) {
            throw error;
        }
    }

    const userId = await findUserId(token, user.usuario);
    if (!userId) {
        throw new Error(`No se pudo obtener el usuario ${user.usuario} en Keycloak`);
    }

    return userId;
}

async function assignRealmRole(token, userId, roleName) {
    const roleResponse = await axios.get(
        `${config.baseUrl}/admin/realms/${config.realm}/roles/${roleName}`,
        { headers: { Authorization: `Bearer ${token}` } }
    );

    await axios.post(
        `${config.baseUrl}/admin/realms/${config.realm}/users/${userId}/role-mappings/realm`,
        [roleResponse.data],
        { headers: { Authorization: `Bearer ${token}` } }
    );
}

async function createUserWithRole(user, roleName) {
    if (!user.usuario || !user.pass) {
        throw new Error('Usuario y pass son obligatorios para crear el usuario en Keycloak');
    }

    const token = await getAdminToken();
    const userId = await createOrGetUser(token, user);
    await assignRealmRole(token, userId, roleName);

    return { userId, roleName };
}

module.exports = {
    createUserWithRole
};

File: backend/m3/backend_flask/src/keycloak/keycloak_admin.py
import json
import os
import urllib.parse
import urllib.request


BASE_URL = os.getenv('KEYCLOAK_INTERNAL_URL') or os.getenv('KEYCLOAK_URL') or 'http://keycloak:8080'
REALM = os.getenv('KEYCLOAK_REALM', 'Universidad')
ADMIN_REALM = os.getenv('KEYCLOAK_ADMIN_REALM', 'master')
ADMIN_USER = os.getenv('KEYCLOAK_ADMIN_USER', 'admin')
ADMIN_PASSWORD = os.getenv('KEYCLOAK_ADMIN_PASSWORD', '123456')
ADMIN_CLIENT = os.getenv('KEYCLOAK_ADMIN_CLIENT', 'admin-cli')


def create_user_with_role(user, role_name):
    if not user.get('usuario') or not user.get('pass'):
        raise ValueError('Usuario y pass son obligatorios para crear el usuario en Keycloak')

    token = get_admin_token()
    user_id = create_or_get_user(token, user)
    assign_realm_role(token, user_id, role_name)
    return {'userId': user_id, 'roleName': role_name}


def get_admin_token():
    data = urllib.parse.urlencode({
        'grant_type': 'password',
        'client_id': ADMIN_CLIENT,
        'username': ADMIN_USER,
        'password': ADMIN_PASSWORD
    }).encode('utf-8')

    response = request(
        f'{BASE_URL}/realms/{ADMIN_REALM}/protocol/openid-connect/token',
        method='POST',
        data=data,
        headers={'Content-Type': 'application/x-www-form-urlencoded'}
    )
    return response['access_token']


def create_or_get_user(token, user):
    payload = {
        'username': user.get('usuario'),
        'firstName': user.get('nombre'),
        'email': user.get('correo') or None,
        'enabled': True,
        'emailVerified': True,
        'credentials': [{
            'type': 'password',
            'value': user.get('pass'),
            'temporary': False
        }]
    }

    try:
        request(
            f'{BASE_URL}/admin/realms/{REALM}/users',
            method='POST',
            token=token,
            data=json.dumps(payload).encode('utf-8')
        )
    except urllib.error.HTTPError as error:
        if error.code != 409:
            raise

    users = request(
        f'{BASE_URL}/admin/realms/{REALM}/users?username={urllib.parse.quote(user.get("usuario"))}&exact=true',
        token=token
    )
    if not users:
        raise RuntimeError(f'No se pudo obtener el usuario {user.get("usuario")} en Keycloak')
    return users[0]['id']


def assign_realm_role(token, user_id, role_name):
    role = request(
        f'{BASE_URL}/admin/realms/{REALM}/roles/{urllib.parse.quote(role_name)}',
        token=token
    )
    request(
        f'{BASE_URL}/admin/realms/{REALM}/users/{user_id}/role-mappings/realm',
        method='POST',
        token=token,
        data=json.dumps([role]).encode('utf-8')
    )


def request(url, method='GET', token=None, data=None, headers=None):
    request_headers = {'Content-Type': 'application/json', **(headers or {})}
    if token:
        request_headers['Authorization'] = f'Bearer {token}'

    req = urllib.request.Request(url, data=data, headers=request_headers, method=method)
    with urllib.request.urlopen(req, timeout=15) as response:
        body = response.read().decode('utf-8')
        return json.loads(body) if body else None
